subsys/net/lib/dhcpv4/dhcpv4.c

```diff
@@ -667,6 +671,56 @@ static void dhcpv4_timeout(struct k_work *work)
 	}
 }
 
+#if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC)
+
+static int dhcpv4_parse_option_vendor(struct net_pkt *pkt, struct net_if *iface,
+				      enum net_dhcpv4_msg_type *msg_type, uint8_t length)
+{
+	struct net_dhcpv4_option_callback *vs_cb, *tmp;
+	struct net_pkt_cursor vs_backup;
+	uint8_t len;
+	int len_counter = length;
+	uint8_t vs_type;
+
+	while (!net_pkt_read_u8(pkt, &vs_type) && len_counter > 0) {
+		if (vs_type == DHCPV4_OPTIONS_END) {
+			NET_DBG("Vendor-specific options_end");
+			return 0;
+		}
+		len_counter--;
+
+		if (net_pkt_read_u8(pkt, &len)) {
+			NET_ERR("Vendor-specific option parsing, bad length");
+			return -ENOBUFS;
+		}
+		len_counter--;
+		if (len_counter < len) {
+			NET_ERR("Vendor-specific option parsing, length too long");
+			return -EBADMSG;
+		}
+		net_pkt_cursor_backup(pkt, &vs_backup);
+
+		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&option_vendor_callbacks, vs_cb, tmp, node) {
+			if (vs_cb->option == vs_type) {
+				NET_ASSERT(vs_cb->handler, "No callback handler!");
+
+				if (net_pkt_read(pkt, vs_cb->data, MIN(vs_cb->max_length, len))) {
+					NET_DBG("option vendor callback, read err");
+					return -ENOBUFS;
+				}
+
+				vs_cb->handler(vs_cb, len, *msg_type, iface);
+				net_pkt_cursor_restore(pkt, &vs_backup);
+			}
+		}
+		net_pkt_skip(pkt, len);
+		len_counter = len_counter - len;
+	}
+	return 0;
```

Comment:
Same here, if we exit here, it means we did not read END marker, IMO that's an error condition.

Reply:
According to the RFC 2132 the existence of the end maker in encapsulated vendor options is not mandatory 
`3) Code 255 (END), if present, signifies the end of the
         encapsulated vendor extensions, not the end of the vendor
         extensions field. If no code 255 is present, then the end of
         the enclosing vendor-specific information field is taken as the
         end of the encapsulated vendor-specific extensions field.`

Reply:
Thanks for clarifying, I got confused as END marker is mandatory in the DHCP option list, but apparently that's no the case for the option sub-list.